runtime/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt

```diff
@@ -80,19 +88,32 @@ public open class PluginGeneratedSerialDescriptor(
         @Suppress("DEPRECATION_ERROR")
         if (other !is SerialDescriptor) return false
         if (serialName != other.serialName) return false
-        // TODO compare only serial names
-        if (elementDescriptors() != other.elementDescriptors()) return false
+        if (typeParameterDescriptors != other.typeParameters()) return false
         return true
     }
 
     override fun hashCode(): Int {
         var result = serialName.hashCode()
-        // TODO hashcode only for serial name
-        result = 31 * result + elementDescriptors().hashCode()
+        result = 31 * result + typeParameterDescriptors.hashCode()
         return result
     }
 
+    // todo: should type parameters be in serial name?
     override fun toString(): String {
-        return indices.entries.joinToString(", ", "$serialName(", ")") { it.key + ": " + getElementDescriptor(it.value).serialName }
+        return indices.entries.joinToString(", ", "$serialName(", ")") {
+            it.key + ": " + getElementDescriptor(it.value).serialName
+        }
     }
 }
+
+// this function is for comparing user-defined and plugin-generated descriptors.
+// Do we really need it?
+@Suppress("DEPRECATION_ERROR")
+internal fun SerialDescriptor.typeParameters(): List<SerialDescriptor> = when (this) {
```

Comment:
I've tried not to add `public val typeParameterDescriptors` to the interface, but this may lead to the problem where some implementations are not equal or not symmetrically equal (e.g. ListLikeDesc and SerialDescImpl).

Perhaps, this needs to be public with a name like `SerialDescriptor.defaultEquals(other)`, so it can be used in user-defined and our own implementations consistently. However, such a decision requires adding `public val typeParameterDescriptors` to the interface.

Reply:
`internal` is fine for now

Reply:
If different implementations of `SerialDescriptor` are considered to be non-equal to each other, this function is not needed at all